### PR TITLE
Added jenkinsfile which give examples and can be used to test

### PR DESCRIPTION
**Pull request by garethahealy**
#### What is this PR About?
Have added a Jenkinsfile per lib method. Each Jenkinsfile can be used as an example of how the pipeline-lib can be used, as well as to test changes.

Related to:
- https://github.com/redhat-cop/pipeline-library/issues/20

#### How do we test this?
Follow the TESTING.md instructions

cc: @redhat-cop/day-in-the-life

### CODE REVIEW

**Comment by garethahealy** [comment]
**DONT MERGE.**

The below need to be merged first:
- https://github.com/redhat-cop/pipeline-library/pull/52
- https://github.com/redhat-cop/pipeline-library/pull/51
- https://github.com/redhat-cop/pipeline-library/pull/50

As the tests will fail without these changes.

**Comment by garethahealy** [comment]
Tested on OCP3 and OCP4 on opentlc infrastructure.

**Comment by garethahealy** [comment]
@pabrahamsson ; ready for review/merge

**Comment by pabrahamsson** [comment]
I've started reviewing but ran into a few issues that I'm working through. The issues are mostly unrelated to the PR and will need to be fixed elsewhere ([#249](https://github.com/redhat-cop/containers-quickstarts/pull/249) in containers-quickstarts for example).
I'm hoping to finish up my review later today.

File: TESTING.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+oc label secret my-token credential.sync.jenkins.openshift.io=true
+
+oc create secret generic --from-literal=registry=$(oc get is jenkins -n openshift -o jsonpath={.status.dockerImageRepository} | cut -d '/' -f1 | xargs) --from-literal=username=openshift1 --from-literal=token=$(oc whoami --show-token) local-registry-generic
+oc secrets link --for=mount jenkins local-registry-generic
+
+oc create secret docker-registry --docker-server=$(oc get is jenkins -n openshift -o jsonpath={.status.dockerImageRepository} | cut -d '/' -f1 | xargs) --docker-username=openshift1 --docker-password=$(oc whoami --show-token) --docker-email=unused local-registry
+oc secrets link --for=mount jenkins local-registry
+
+### Start all BuildConfigs
+find test -type f -name "Jenkinsfile-*" -exec bash -c 'oc start-build $(basename {} | tr 'A-Z' 'a-z')-pipeline' \;
```
Comment (pabrahamsson) [inline TESTING.md]: Could you add a step about cleaning up as we are creating a few new projects?

File: test/Jenkinsfile-applyTemplate [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+node("maven") {
+    stage("SETUP: Create deployment files") {
+        def params = """
+                    NAME=applytemplate
+                    DATABASE_SERVICE_NAME=applytemplate
+                    """
+
+        writeFile file: "params.txt", text: params
+
+        sh "curl -L -o cakephp-mysql-persistent.json https://raw.githubusercontent.com/openshift/origin/v3.11.0/examples/quickstarts/cakephp-mysql-persistent.json"
```
Comment (pabrahamsson) [inline test/Jenkinsfile-applyTemplate]: Is there a particular reason for using persistence here? Would it suffice with ephemeral version of cake php?

File: test/Jenkinsfile-rollback [hunk 3]
```diff
@@ -0,0 +1,61 @@
+#!groovy
+@Library(["pipeline-library@master"]) _
+
+def firstDeploymentVersion
+
+node("maven") {
+    stage("SETUP: Create deployment files") {
+        openshift.withCluster() {
+            openshift.withProject() {
+                def model = openshift.process("https://raw.githubusercontent.com/openshift/origin/v3.11.0/examples/quickstarts/cakephp-mysql-persistent.json", "-p NAME=rollback", "-p DATABASE_SERVICE_NAME=rollback")
```
Comment (pabrahamsson) [inline test/Jenkinsfile-rollback]: See above regarding persistence

File: test/Jenkinsfile-rollout [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#!groovy
+@Library(["pipeline-library@master"]) _
+
+def firstDeploymentVersion
+
+node("maven") {
+    stage("SETUP: Create deployment files") {
+        openshift.withCluster() {
+            openshift.withProject() {
+                def model = openshift.process("https://raw.githubusercontent.com/openshift/origin/v3.11.0/examples/quickstarts/cakephp-mysql-persistent.json", "-p NAME=rollout", "-p DATABASE_SERVICE_NAME=rollout")
```
Comment (pabrahamsson) [inline test/Jenkinsfile-rollout]: See above regarding persistence

File: test/Jenkinsfile-tagAndDeploy [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            openshift.withCluster() {
+                openshift.withProject("openshift") {
+                    def imageStream = openshift.selector("is", "jenkins")
+                    dockerRegistry = sh(returnStdout: true, script: "echo '${imageStream.object().status.dockerImageRepository}' | cut -d '/' -f1").trim()
+                }
+            }
+
+            openshift.withCluster() {
+                openshift.withProject() {
+                    def model = openshift.process("https://raw.githubusercontent.com/openshift/origin/v3.11.0/examples/quickstarts/cakephp-mysql-persistent.json", "-p NAME=taganddeploy", "-p DATABASE_SERVICE_NAME=taganddeploy")
```
Comment (pabrahamsson) [inline test/Jenkinsfile-tagAndDeploy]: See above regarding persistence

**Comment by garethahealy** [comment]
@pabrahamsson ; pushed requested changes

File: test/Jenkinsfile-applyTemplate [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+node("maven") {
+    stage("SETUP: Create deployment files") {
+        def params = """
+                    NAME=applytemplate
+                    DATABASE_SERVICE_NAME=applytemplate
+                    """
+
+        writeFile file: "params.txt", text: params
+
+        sh "curl -L -o cakephp-mysql-persistent.json https://raw.githubusercontent.com/openshift/origin/v3.11.0/examples/quickstarts/cakephp-mysql.json"
```
Comment (pabrahamsson) [inline test/Jenkinsfile-applyTemplate]: Minor nitpick, can you also change the downloaded filename to `cakephp-mysql.json` to avoid confusing anyone?

**Comment by garethahealy** [comment]
@pabrahamsson ; pushed.

**Comment by garethahealy** [comment]
@pabrahamsson ; once you've merged, can you also do a release?
- https://github.com/redhat-cop/pipeline-library/issues/58

**Pull request merged**